### PR TITLE
Fuse gate_proj and up_proj in Qwen 2.5 VL's vision MLP

### PR DESCRIPTION
**Pull request by AlienKevin**
## Motivation

Fused MLPs are widely used in LLMs like Qwen 2 but was not used in Qwen 2.5 VL's vision module. This PR adds MLP fusion to the vision module to gain a 4.5% boost in vision prefill throughput (10045 -> 10498 input tok/s). We expect more throughput boost when multiple requests are batch encoded in the future.

## Accuracy Tests

MMMU overall accuracy preserved compared to Triton backend before PR (0.513 vs 0.514).
Before PR (0.514):
```
Evaluating...
{'Accounting': {'acc': 0.433, 'num': 30},
 'Agriculture': {'acc': 0.533, 'num': 30},
 'Architecture_and_Engineering': {'acc': 0.5, 'num': 30},
 'Art': {'acc': 0.667, 'num': 30},
 'Art_Theory': {'acc': 0.833, 'num': 30},
 'Basic_Medical_Science': {'acc': 0.6, 'num': 30},
 'Biology': {'acc': 0.367, 'num': 30},
 'Chemistry': {'acc': 0.333, 'num': 30},
 'Clinical_Medicine': {'acc': 0.633, 'num': 30},
 'Computer_Science': {'acc': 0.433, 'num': 30},
 'Design': {'acc': 0.733, 'num': 30},
 'Diagnostics_and_Laboratory_Medicine': {'acc': 0.467, 'num': 30},
 'Economics': {'acc': 0.467, 'num': 30},
 'Electronics': {'acc': 0.267, 'num': 30},
 'Energy_and_Power': {'acc': 0.267, 'num': 30},
 'Finance': {'acc': 0.367, 'num': 30},
 'Geography': {'acc': 0.4, 'num': 30},
 'History': {'acc': 0.667, 'num': 30},
 'Literature': {'acc': 0.767, 'num': 30},
 'Manage': {'acc': 0.367, 'num': 30},
 'Marketing': {'acc': 0.467, 'num': 30},
 'Materials': {'acc': 0.433, 'num': 30},
 'Math': {'acc': 0.5, 'num': 30},
 'Mechanical_Engineering': {'acc': 0.5, 'num': 30},
 'Music': {'acc': 0.433, 'num': 30},
 'Overall': {'acc': 0.514, 'num': 900},
 'Overall-Art and Design': {'acc': 0.667, 'num': 120},
 'Overall-Business': {'acc': 0.42, 'num': 150},
 'Overall-Health and Medicine': {'acc': 0.593, 'num': 150},
 'Overall-Humanities and Social Science': {'acc': 0.667, 'num': 120},
 'Overall-Science': {'acc': 0.42, 'num': 150},
 'Overall-Tech and Engineering': {'acc': 0.419, 'num': 210},
 'Pharmacy': {'acc': 0.633, 'num': 30},
 'Physics': {'acc': 0.5, 'num': 30},
 'Psychology': {'acc': 0.7, 'num': 30},
 'Public_Health': {'acc': 0.633, 'num': 30},
 'Sociology': {'acc': 0.533, 'num': 30}}
eval out saved to ./val_sglang.json
Overall accuracy: 0.514
```

After PR (0.513):
```
{'Accounting': {'acc': 0.433, 'num': 30},
 'Agriculture': {'acc': 0.533, 'num': 30},
 'Architecture_and_Engineering': {'acc': 0.467, 'num': 30},
 'Art': {'acc': 0.667, 'num': 30},
 'Art_Theory': {'acc': 0.8, 'num': 30},
 'Basic_Medical_Science': {'acc': 0.6, 'num': 30},
 'Biology': {'acc': 0.367, 'num': 30},
 'Chemistry': {'acc': 0.333, 'num': 30},
 'Clinical_Medicine': {'acc': 0.633, 'num': 30},
 'Computer_Science': {'acc': 0.367, 'num': 30},
 'Design': {'acc': 0.7, 'num': 30},
 'Diagnostics_and_Laboratory_Medicine': {'acc': 0.467, 'num': 30},
 'Economics': {'acc': 0.5, 'num': 30},
 'Electronics': {'acc': 0.3, 'num': 30},
 'Energy_and_Power': {'acc': 0.267, 'num': 30},
 'Finance': {'acc': 0.333, 'num': 30},
 'Geography': {'acc': 0.4, 'num': 30},
 'History': {'acc': 0.667, 'num': 30},
 'Literature': {'acc': 0.833, 'num': 30},
 'Manage': {'acc': 0.367, 'num': 30},
 'Marketing': {'acc': 0.5, 'num': 30},
 'Materials': {'acc': 0.433, 'num': 30},
 'Math': {'acc': 0.467, 'num': 30},
 'Mechanical_Engineering': {'acc': 0.5, 'num': 30},
 'Music': {'acc': 0.5, 'num': 30},
 'Overall': {'acc': 0.513, 'num': 900},
 'Overall-Art and Design': {'acc': 0.667, 'num': 120},
 'Overall-Business': {'acc': 0.427, 'num': 150},
 'Overall-Health and Medicine': {'acc': 0.593, 'num': 150},
 'Overall-Humanities and Social Science': {'acc': 0.683, 'num': 120},
 'Overall-Science': {'acc': 0.407, 'num': 150},
 'Overall-Tech and Engineering': {'acc': 0.41, 'num': 210},
 'Pharmacy': {'acc': 0.633, 'num': 30},
 'Physics': {'acc': 0.467, 'num': 30},
 'Psychology': {'acc': 0.7, 'num': 30},
 'Public_Health': {'acc': 0.633, 'num': 30},
 'Sociology': {'acc': 0.533, 'num': 30}}
eval out saved to ./val_sglang.json
Overall accuracy: 0.513
```

## Benchmarking and Profiling

4.5% boost in vision prefill throughput (10045 -> 10498 input tok/s).

Before MLP fusion:
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 not set   
Successful requests:                     768       
Benchmark duration (s):                  78.21     
Total input tokens:                      785664    
Total generated tokens:                  768       
Total generated tokens (retokenized):    768       
Request throughput (req/s):              9.82      
Input token throughput (tok/s):          10045.23  
Output token throughput (tok/s):         9.82      
Total token throughput (tok/s):          10055.05  
Concurrency:                             512.55    
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   52197.37  
Median E2E Latency (ms):                 52393.05  
---------------Time to First Token----------------
Mean TTFT (ms):                          52197.35  
Median TTFT (ms):                        52393.04  
P99 TTFT (ms):                           77564.23  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

After fusion:
```
============ Serving Benchmark Result ============
Backend:                                 sglang    
Traffic request rate:                    inf       
Max request concurrency:                 not set   
Successful requests:                     768       
Benchmark duration (s):                  74.84     
Total input tokens:                      785664    
Total generated tokens:                  768       
Total generated tokens (retokenized):    768       
Request throughput (req/s):              10.26     
Input token throughput (tok/s):          10498.03  
Output token throughput (tok/s):         10.26     
Total token throughput (tok/s):          10508.29  
Concurrency:                             527.95    
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   51447.35  
Median E2E Latency (ms):                 51471.03  
---------------Time to First Token----------------
Mean TTFT (ms):                          51447.33  
Median TTFT (ms):                        51471.02  
P99 TTFT (ms):                           74112.44  
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/models/qwen2_5_vl.py [hunk 1]
```diff
@@ -455,8 +449,8 @@ class Qwen2_5_VLForConditionalGeneration(nn.Module):
         "q_proj": ("qkv_proj", 0),
         "k_proj": ("qkv_proj", 1),
         "v_proj": ("qkv_proj", 2),
-        "gate_proj": ("gate_up_proj", 0),
-        "up_proj": ("gate_up_proj", 1),
+        "gate_up_proj": ("gate_up_proj", 0),
+        "down_proj": ("gate_up_proj", 1),
```
[the remark this answers is not in the document]
Comment (JustinTong0323) [reply python/sglang/srt/models/qwen2_5_vl.py]: @AlienKevin I am not familiar with bitsandbytes quantization, have you tested on it?

[the remark this answers is not in the document]
Comment (AlienKevin) [reply python/sglang/srt/models/qwen2_5_vl.py]: @JustinTong0323 I'm also not familiar with bitsandbytes and haven't tested it. Looking through the code again I believe Gemini is right: the keys should be the original unfused param names. Applied Gemini's fix.

**Review by JustinTong0323 — APPROVED** [review]
Thanks Kevin!!!

**Comment by JustinTong0323** [comment]
Amazing latency&throughout improvement on benchmark:

Launch cmd:
`python -m sglang.launch_server --model-path Qwen/Qwen2.5-VL-3B-Instruct --disable-radix-cache`

Bench cmd:
```
python -m sglang.bench_serving \
    --backend sglang-oai-chat \
    --dataset-name random-image \
    --num-prompts 100 \
    --random-image-num-images 3 \
    --random-image-resolution 1080p \
    --random-input-len 512 \
    --random-output-len 512 --port 30000
```


Before this PR:
```
Input token throughput (tok/s):          293.38    
Output token throughput (tok/s):         272.62    
Mean ITL (ms):                           625.21    
```
After this PR:
```
Input token throughput (tok/s):          329.35    
Output token throughput (tok/s):         308.63 
Mean ITL (ms):                           493.41    
```
